bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/Workbench.java

```diff
@@ -1483,6 +1489,63 @@ IWorkbenchWindow createWorkbenchWindow(IAdaptable input, IPerspectiveDescriptor
 		return result;
 	}
 
+	private void placeNearActiveShell(MWindow window) {
+		if (getDisplay() == null)
+			return;
+
+		Shell activeShell = getDisplay().getActiveShell();
+		if (activeShell == null)
+			return;
+
+		Rectangle newShellBounds = activeShell.getBounds();
+		Monitor currentMonitor = findMonitorThatContains(newShellBounds.x, newShellBounds.y);
+
+		Rectangle currentMonitorBounds = currentMonitor.getBounds();
+
+		final int offsetToExistingShell = 100;
+		newShellBounds.x += offsetToExistingShell;
+		newShellBounds.y += offsetToExistingShell;
+		moveIntoBounds(newShellBounds, currentMonitorBounds);
+		window.setX(newShellBounds.x);
+		window.setY(newShellBounds.y);
+	}
+
+	/**
+	 * Find the monitor that contains a point.
+	 *
+	 * @param x the coordinate "X"
+	 * @param y the coordinate "Y"
+	 * @return the first monitor that contains the point or, if none is found, the
+	 *         primary monitor.
+	 */
+	private Monitor findMonitorThatContains(int x, int y) {
+		for (Monitor m : getDisplay().getMonitors()) {
+			if (m.getBounds().contains(x, y))
+				return m;
+		}
+
+		return getDisplay().getPrimaryMonitor();
+	}
+
+	private static void moveIntoBounds(Rectangle rectangleToMove, Rectangle bounds) {
+		// move into bounds if it's too far to the right
+		if (rectangleToMove.x + rectangleToMove.width > bounds.x + bounds.width)
+			rectangleToMove.x = bounds.x + bounds.width - rectangleToMove.width - 100;
+
+		// move into bounds if it's too far to the left
+		if (rectangleToMove.x < bounds.x)
+			rectangleToMove.x = bounds.x;
```

Comment:
Isn't this dead code? The display bounds x and y values are by construction lower than the values of the new shell anyway.

Reply:
Nope, if the current shell is too far to the left (_e.g._ x = -300) then `rectangleToMove` will also be outside the display bounds. Same thing with the y-coordinate.

It should not happen often but it could happen and if it does, I find it helpful to at least have the upper-left corner of the window inside the monitor so the user can grab and move the window around.

Reply:
I see, thanks for pointing that out. I don't even think that it is too unlikly to happen, so this correction is necessary.

But this points me to some other behavior I would not expect: In case the current workbench window is placed left of the leftmost monitor (i.e., `x < 0`), then the new window will be placed on the primary monitor rather than on the leftmost one. I would expect the new workbench to placed on the leftmost monitor in that situation. So technically, I would expect `x` and `y` to be set to zero if they are negative before calculating any assignment to a monitor.
As a consequence, handling negative values would not necessary at this place anymore.

Reply:
Good point, that was a bug in the method `findMonitor`. I corrected it in 2521dd5850586f9c75d833114f9abab937019c15 by replacing the implementation to look for the monitor that holds the biggest portion of the current window. The default monitor is still used as back up in case the current window is outside all monitors... if that's even possible

Reply:
I like that solution. It's probably closest to the behavior a user might expect.